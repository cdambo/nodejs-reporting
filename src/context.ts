export default class Context {
  private readonly globalTags: object;

  public constructor(globalTags: object = {}) {
    this.globalTags = globalTags;
  }

  public mergeTags(tags: object = {}): object {
    return Object.assign({}, this.globalTags, tags);
  }
}
