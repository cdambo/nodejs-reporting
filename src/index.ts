import Context from "./context";

export * from "./metrics";
export { Context };
