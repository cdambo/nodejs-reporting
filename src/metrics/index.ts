export * from "./formats";
export * from "./reporters";
export * from "./integrations";
