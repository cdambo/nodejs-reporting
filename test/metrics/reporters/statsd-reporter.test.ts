import { StatsD } from "node-dogstatsd";
import { StatsDReporter, StatsDArgs, StatsDFormat } from "../../../src";

const getReporter = (): StatsDReporter =>
  new StatsDReporter({
    host: "host",
    globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
  });

describe("StatsDReporter", (): void => {
  describe("Constructor", (): void => {
    it("Creates a StatsDReporter", (): void => {
      const reporter = new StatsDReporter();
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a StatsDReporter with global tags", (): void => {
      const reporter = new StatsDReporter({
        globalTags: { globalTag1: "global_one", globalTag2: "global_two" }
      });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a StatsDReporter with a client", (): void => {
      const reporter = new StatsDReporter({
        client: new StatsD("host", 8888, null)
      });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a StatsDReporter with a host", (): void => {
      const reporter = new StatsDReporter({ host: "host" });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a StatsDReporter with a port", (): void => {
      const reporter = new StatsDReporter({ port: 5555 });
      expect(reporter).toMatchSnapshot();
    });

    it("Creates a StatsDReporter with a format", (): void => {
      const format: StatsDFormat = {
        format: (): StatsDArgs => ({ stat: "stat" })
      };
      const reporter = new StatsDReporter({ format });
      expect(reporter).toMatchSnapshot();
    });
  });

  describe("timing", (): void => {
    it("Reports timing metric", (): void => {
      const reporter = getReporter();
      const mockTiming = jest
        .spyOn(reporter.client, "timing")
        .mockImplementation();
      reporter.timing("timing_test", 20, 0.5, { tag1: "one", tag2: "two" });
      expect(mockTiming.mock.calls).toMatchSnapshot();
    });
  });

  describe("increment", (): void => {
    it("Reports counter metric", (): void => {
      const reporter = getReporter();
      const mockIncrement = jest
        .spyOn(reporter.client, "increment")
        .mockImplementation();
      reporter.increment("increment_test", 0.5, { tag1: "one", tag2: "two" });
      expect(mockIncrement.mock.calls).toMatchSnapshot();
    });
  });

  describe("incrementBy", (): void => {
    it("Reports counter metric", (): void => {
      const reporter = getReporter();
      const mockIncrementBy = jest
        .spyOn(reporter.client, "incrementBy")
        .mockImplementation();
      reporter.incrementBy("increment_by_test", 10, {
        tag1: "one",
        tag2: "two"
      });
      expect(mockIncrementBy.mock.calls).toMatchSnapshot();
    });
  });

  describe("decrement", (): void => {
    it("Reports counter metric", (): void => {
      const reporter = getReporter();
      const mockDecrement = jest
        .spyOn(reporter.client, "decrement")
        .mockImplementation();
      reporter.decrement("decrement_test", 0.5, { tag1: "one", tag2: "two" });
      expect(mockDecrement.mock.calls).toMatchSnapshot();
    });
  });

  describe("decrementBy", (): void => {
    it("Reports counter metric", (): void => {
      const reporter = getReporter();
      const mockDecrementBy = jest
        .spyOn(reporter.client, "decrementBy")
        .mockImplementation();
      reporter.decrementBy("decrement_by_test", 10, {
        tag1: "one",
        tag2: "two"
      });
      expect(mockDecrementBy.mock.calls).toMatchSnapshot();
    });
  });

  describe("gauge", (): void => {
    it("Reports gauge metric", (): void => {
      const reporter = getReporter();
      const mockGauge = jest
        .spyOn(reporter.client, "gauge")
        .mockImplementation();
      reporter.gauge("gauge_test", 10, 0.5, { tag1: "one", tag2: "two" });
      expect(mockGauge.mock.calls).toMatchSnapshot();
    });
  });

  describe("histogram", (): void => {
    it("Reports histogram metric", (): void => {
      const reporter = getReporter();
      const mockHistogram = jest
        .spyOn(reporter.client, "histogram")
        .mockImplementation();
      reporter.histogram("histogram_test", 20, 0.5, {
        tag1: "one",
        tag2: "two"
      });
      expect(mockHistogram.mock.calls).toMatchSnapshot();
    });
  });

  describe("close", (): void => {
    it("Closes the socket", (): void => {
      const reporter = getReporter();
      const mockClose = jest
        .spyOn(reporter.client, "close")
        .mockImplementation();
      reporter.close();
      expect(mockClose.mock.calls).toMatchSnapshot();
    });
  });
});
