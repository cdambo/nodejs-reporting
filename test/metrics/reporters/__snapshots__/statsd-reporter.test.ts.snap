// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsDReporter Constructor Creates a StatsDReporter 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "localhost",
    "last_used_timer": null,
    "port": 8125,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter Constructor Creates a StatsDReporter with a client 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "host",
    "last_used_timer": null,
    "port": 8888,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter Constructor Creates a StatsDReporter with a format 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "localhost",
    "last_used_timer": null,
    "port": 8125,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter Constructor Creates a StatsDReporter with a host 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "host",
    "last_used_timer": null,
    "port": 8125,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter Constructor Creates a StatsDReporter with a port 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "localhost",
    "last_used_timer": null,
    "port": 5555,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {},
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter Constructor Creates a StatsDReporter with global tags 1`] = `
StatsDReporter {
  "client": Client {
    "ephemeral_socket": null,
    "global_tags": undefined,
    "host": "localhost",
    "last_used_timer": null,
    "port": 8125,
    "socket": null,
  },
  "context": Context {
    "globalTags": Object {
      "globalTag1": "global_one",
      "globalTag2": "global_two",
    },
  },
  "format": Object {
    "format": [Function],
  },
}
`;

exports[`StatsDReporter close Closes the socket 1`] = `
Array [
  Array [],
]
`;

exports[`StatsDReporter decrement Reports a counter metric 1`] = `
Array [
  Array [
    "decrement_test",
    0.5,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter decrementBy Reports a counter metric 1`] = `
Array [
  Array [
    "decrement_by_test",
    10,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter gauge Reports a gauge metric 1`] = `
Array [
  Array [
    "gauge_test",
    10,
    0.5,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter histogram Reports a histogram metric 1`] = `
Array [
  Array [
    "histogram_test",
    20,
    0.5,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter increment Reports a counter metric 1`] = `
Array [
  Array [
    "increment_test",
    0.5,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter incrementBy Reports a counter metric 1`] = `
Array [
  Array [
    "increment_by_test",
    10,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;

exports[`StatsDReporter timing Reports a timing metric 1`] = `
Array [
  Array [
    "timing_test",
    20,
    0.5,
    Array [
      "tag_1:one",
      "tag_2:two",
    ],
  ],
]
`;
